src/Views.Win32/components/ActionMenu.cpp

```diff
@@ -0,0 +1,380 @@
+﻿/*
+ * Copyright (c) 2025, Mupen64 maintainers, contributors, and original authors (Hacktarux, ShadowPrince, linker).
+ *
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ */
+
+#include "stdafx.h"
+#include <ActionManager.h>
+#include <Messenger.h>
+#include <components/ActionMenu.h>
+
+const auto MANAGED_MENU_CTX = L"Mupen64_ManagedMenuContext";
+
+struct t_menu_item {
+    uint16_t id{};
+    size_t position_under_parent{};
+    HMENU popup_handle{};
+    HMENU parent_menu{};
+    bool has_menu{};
+
+    std::wstring action_path{};
+    std::vector<t_menu_item> children{};
+
+private:
+    std::wstring m_path{};
+    bool m_has_separator{};
+
+public:
+    explicit t_menu_item(const std::wstring& path);
+
+    /**
+     * \brief Performs a depth-first iteration over the menu item tree, applying the given predicate to each item. The predicate is also applied to the initial item itself.
+     */
+    void iterate_children_and_self(const std::function<void(t_menu_item& item)>& predicate);
+
+    [[nodiscard]] auto raw_path() const
+    {
+        return m_path;
+    }
+
+    [[nodiscard]] bool has_separator() const
+    {
+        return m_has_separator;
+    }
+};
+
+struct t_action_menu_context {
+    HWND hwnd{};
+    t_menu_item menu{L"Root"};
+    size_t menu_id_counter{};
+};
+
+struct t_action_menu_global_context {
+    std::vector<t_action_menu_context*> active_contexts{};
+    std::vector<std::wstring> actions{};
+};
+
+static t_action_menu_global_context g_am_ctx{};
+
+t_menu_item::t_menu_item(const std::wstring& path)
+{
+    this->m_path = path;
+
+    const auto name = ActionManager::get_segments(path).back();
+    this->m_has_separator = name.ends_with(ActionManager::SEPARATOR_SUFFIX);
+}
+
+void t_menu_item::iterate_children_and_self(const std::function<void(t_menu_item& item)>& predicate)
+{
+    predicate(*this);
+    for (auto& child : children)
+    {
+        child.iterate_children_and_self(predicate);
+    }
+}
```

Comment:
The `predicate` is a complete misnomer here, because a predicate is (almost?) always a pure function that returns a boolean value for an element of some set.  
This here, however, is an action to perform on every hierarchically lower or equal element as `this` itself.  
I suggest simply calling the argument `action`.
```suggestion
void t_menu_item::iterate_children_and_self(const std::function<void(t_menu_item& item)>& action)
{
    action(*this);
    for (auto& child : children)
    {
        child.iterate_children_and_self(action);
    }
}
```

Reply:
Hello from C# land with Action<T>? :P